clients/src/main/java/org/apache/kafka/common/security/auth/KafkaPrincipal.java

```diff
@@ -68,9 +73,15 @@ public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null) return false;
         if (getClass() != o.getClass()) return false;
-
         KafkaPrincipal that = (KafkaPrincipal) o;
-        return principalType.equals(that.principalType) && name.equals(that.name);
+
+        // hashes are cached, so this goes quicker than the string compare
+        if (this.hashCode() != o.hashCode()) {
```

Comment:
This is incorrect - hashCode can have collisions.

Reply:
You  are correct that they can collide, but if the hashCode of 2 strings don't match then they are for sure not the same.
I.e. 
```
A.hashcode() != B.hashcode() == true => they are not the same 
A.hashcode() != B.hashcode() == false => they are the same OR they collide

//Hence the check after this to do the value compare:
return name.equals(that.name) && principalType.equals(that.principalType);
```